lib/cisco_node_utils/cmd_ref/README_YAML.md

```diff
@@ -0,0 +1,458 @@
+`_append` variants) all support two forms of wildcarding - printf-style and
+key-value. Key-value is generally preferred, as described below.
+
+#### Printf-style wildcards
+
+```yaml
+# tacacs_server_host.yaml
+encryption:
+  config_set: '%s tacacs-server host %s key %s %s'
+```
+
+This permits parameter values to be passed as a simple sequence to generate the resulting string or regexp:
+
+```ruby
+irb(main):009:0> ref = cr.lookup('tacacs_server_host', 'encryption')
+irb(main):010:0> ref.config_set('no', 'myhost', 'md5', 'mypassword')
+=> ["no tacacs-server host myhost key md5 mypassword"]
+```
+
+This approach is quick to implement and concise, but less flexible - in
+particular it cannot handle a case where different platforms (or different
+client APIs!) take parameters in a different order - and less readable in
+the Ruby code.
+
+#### Key-value wildcards
+
+```yaml
+# ospf.yaml
+auto_cost:
+  config_set: ['router ospf <name>', 'auto-cost reference-bandwidth <cost> <type>']
+```
+
+This requires parameter values to be passed as a hash:
+
+```ruby
+irb(main):015:0> ref = cr.lookup('ospf', 'auto_cost')
+irb(main):016:0> ref.config_set(name: 'red', cost: '40', type: 'Gbps')
+=> ["router ospf red", "auto-cost reference-bandwidth 40 Gbps"]
+```
+
+This approach is moderately more complex to implement but is more readable in
+the ruby code and is flexible enough to handle significant platform
+differences in CLI. It is therefore the recommended approach for new
+development.
+
+## Advanced attribute definition
+
+### `_template`
+
+The optional `_template` section can be used to define base parameters for all
+attributes of a given feature. For example, all interface attributes might be
+checked with the `show running-config interface all` command, and all
+attributes might be set by first entering the interface configuration submode
+with the `interface <name>` configuration command. Thus, you might have:
+
+```yaml
+# interface.yaml
+_template:
+  config_get: 'show running-config interface all'
+  config_get_token: '/^interface <name>$/'
+  config_set: 'interface <name>'
+
+access_vlan:
+  config_get_token_append: '/^switchport access vlan (.*)$/'
+  config_set_append: 'switchport access vlan <number>'
+
+description:
+  config_get_token_append: '/^description (.*)$/'
+  config_set_append: 'description <desc>'
+
+...
+```
+
+instead of the more repetitive (but equally valid):
+
+```yaml
+# interface.yaml
+access_vlan:
+  config_get: 'show running interface all'
+  config_get_token: ['/^interface %s$/i', '/^switchport access vlan (.*)$/']
+  config_set: ['interface %s', 'switchport access vlan %s']
+
+description:
+  config_get: 'show running-config interface all'
+  config_get_token: ['/^interface <name>$/i', '/^description (.*)$/']
+  config_set: ['interface <name>', 'description <desc>']
+
+...
+```
+
+### API variants
+
+Any of the attribute properties can be subdivided by client API using the
```

Comment:
I feel like this needs more context.  Can we add a reference pointer that describes the client API options in more detail?